spec/services/tax_tribs/case_details_pdf_spec.rb

```diff
@@ -0,0 +1,110 @@
+require 'spec_helper'
+
+RSpec.describe TaxTribs::CaseDetailsPdf do
+  let(:tribunal_case) {
+    TribunalCase.create(
+      case_reference: 'TC/2016/12345',
+      files_collection_ref: 'd29210a8-f2fe-4d6f-ac96-ea4f9fd66687',
+      taxpayer_individual_first_name: 'Firstname',
+      taxpayer_individual_last_name: 'Lastname'
+    )
+  }
+  let(:controller_ctx) {
+    double(
+      ApplicationController,
+      pdf_template: 'show_this',
+      render_to_string: 'this is a pdf',
+      current_tribunal_case: tribunal_case
+    )
+  }
+  let(:presenter) { instance_double(CheckAnswers::AnswersPresenter, pdf_filename: 'TC_2016_12345_FirstnameLastname') }
+  let(:uploader_double) { instance_double(DocumentUpload, upload!: true) }
+
+  subject { described_class.new(tribunal_case, controller_ctx, presenter) }
+
+  describe '#generate' do
+    it 'should generate the PDF' do
+      expect(controller_ctx).to receive(:render_to_string).with(
+        formats: [:pdf],
+        pdf: true,
+        encoding: 'UTF-8',
+        template: 'show_this'
+      )
+      subject.generate
+      expect(subject.pdf).not_to be_nil
+    end
+  end
+
+  describe '#generate_and_upload' do
+    before do
+      allow(DocumentUpload).to receive(:new).and_return(uploader_double)
+      allow(uploader_double).to receive(:valid?).and_return(true)
+      allow(uploader_double).to receive(:errors?).and_return(false)
+    end
+
+    it 'calls generate' do
+      expect(subject).to receive(:generate).and_call_original
```

Comment:
In fact, the whole `before` block could go as it is stubbing out things in `.upload`.